pypif_sdk/func/calculate_funcs.py

```diff
@@ -0,0 +1,280 @@
+                items = (items[0], 1)
+        this_e = {"symbol": items[0], "occurances": float(items[1])}
+        
+        split_element_array.append(this_e)
+    return split_element_array
+
+def _consolidate_elemental_array_(elemental_array):
+    """
+    Accounts for non-empirical chemical formulas by taking in the compositional array generated by _create_compositional_array_() and returning a consolidated array of dictionaries with no repeating elements
+
+    :param elemental_array is an elemental array generated from _create_compositional_array_()
+
+    Returns an array of element dictionaries
+    """
+    condensed_array = []
+    for e in elemental_array:
+        exists = False
+        for k in condensed_array:
+            if k["symbol"] == e["symbol"]:
+                exists = True
+                k["occurances"] += e["occurances"]
+        if not exists:
+            condensed_array.append(e)
+    return condensed_array
+
+def _calculate_n_atoms_(elemental_array):
+    """
+    Calculates the total number of atoms in the system.
+    :param elemental_array is an array of dictionaries as generated by either _create_compositional_array_() or _consolidate_elemental_array_()
+    returns the total number of atoms (can be a partial number in the case of alloys or partial hydrates)
+    """
+    n = 0
+    for a in elemental_array:
+        n += a["occurances"]
+    return n
+
+def _add_ideal_atomic_weights_(elemental_array):
+    """
+    Uses mendeleev.element to find the molar mass of the element in question, and then multiplies that by the occurances of the element.
+    Adds the "weight" property to each of the dictionaries in elemental_array
+    :param elemental_array an array of dictionaries containing information about the elements in the system 
+    Returns the appended elemental_array
+    """
+    for a in elemental_array:
+        # this_atomic_weight = element(a["symbol"]).atomic_weight
+        this_atomic_weight = elements_data[a["symbol"]]["atomic_weight"]
+        a["weight"] = round(a["occurances"] * this_atomic_weight, 2)
+    return elemental_array
+
+def _calculate_total_mass_(elemental_array):
+    """
+    Sums the total weight from all the element dictionaries in the elemental_array after each weight has been added (after _add_ideal_atomic_weights_())
+    :param elemental_array an array of dictionaries containing information about the elements in the system 
+    """
+    n = 0
+    for a in elemental_array:
+        n += a["weight"]
+    return n
+
+def _add_ideal_weight_percent_(elemental_array):
+    """
+    Adds the "weight_percent" property to each of the dictionaries in elemental_array
+    :param elemental_array an array of dictionaries containing information about the elements in the system 
+    Returns the appended elemental_array
+    """
+    t_mass = _calculate_total_mass_(elemental_array)
+    for a in elemental_array:
+        a["weight_percent"] = round(a["weight"]/t_mass * 100, 2)
```

Comment:
Why round? That makes the weight percent composition non-equivalent to the original chemical formula. The atomic percents are real-valued, anyways.